services/cd-service/pkg/repository/db.go

```diff
@@ -165,8 +166,34 @@ func PostgresToSqliteQuery(query string) string {
 	return q
 }
 
+type DBFunction func(ctx context.Context) error
+
+// WithTransaction opens a transaction, runs `f` and then calls either Commit or Rollback
+func (h *DBHandler) WithTransaction(ctx context.Context, f DBFunction) error {
+	tx, err := h.DB.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer func(tx *sql.Tx) {
+		err := tx.Rollback()
+		if err != nil {
+			logger.FromContext(ctx).Warn("db.rollback.failed", zap.Error(err))
+		}
+	}(tx)
+
+	err = f(ctx)
+	if err != nil {
```

Comment:
So there are two errors that can happen in one call of WithTransaction:
* one from either f() or Commit()
* one from Rollback()
The latter we mostly ignore. I wonder if we should wrap both in a custom error type containing both errors -- but discarded this as likely overengineered, because even the calling code will unlikely be able to recover from a failed Rollback. However, best it might do is decide to panic, if this was a critical op as the DB will be inconsistent, which isnt much help. => fine with this.

Reply:
yeah,  I think there's not much we can do to handle a failing Rollback, hence the log. We should consider adding a metric on this, I'm just not quite convinced yet that it's helpful.